gems/aws-sdk-core/lib/aws-sdk-core/ec2_metadata.rb

```diff
@@ -0,0 +1,147 @@
+require 'time'
+require 'net/http'
+
+module Aws
+  class EC2Metadata
+    # Path for PUT request for token
+    # @api private
+    METADATA_TOKEN_PATH = '/latest/api/token'.freeze
+
+    # @api private
+    class TokenRetrievalError < RuntimeError; end
+
+    # @api private
+    class TokenExpiredError < RuntimeError; end
+
+    # @api private
+    class MetadataNotFoundError < RuntimeError; end
+
+    # @api private
+    class RequestForbiddenError < RuntimeError; end
+
+    # TODO - docs
+    def initialize(options = {})
+      @retries = options[:retries] || 1
+      @endpoint = options[:endpoint] || '169.254.169.254'
+      @port = options[:port] || 80
+      @http_open_timeout = options[:http_open_timeout] || 1
+      @http_read_timeout = options[:http_read_timeout] || 1
+      @http_debug_output = options[:http_debug_output]
+      @backoff = backoff(options[:backoff])
```

Comment:
Given our overall basic implantation and that this should all be local, I'd lean towards removing backoff (as a parameter and as a behavior).

Reply:
The IMDS team was keen on having backoff. I'll forward you that context.